src/c/types/basic.c

```diff
@@ -21,133 +21,206 @@
+            { \
+                *value = (TYPE)*us->iterator; \
+                ucdr_advance_stream(us, 1); \
+                serialized = true; \
+            } \
+            else \
+            { \
+                ucdr_promote_buffer(us); \
+            } \
+        } while (!serialized); \
+    } \
+    else \
+    { \
+        us->error = true; \
     } \
-    return !ub->error;
-
-#define UCDR_DESERIALIZE_BYTE_2_CORE() \
-    uint8_t* bytes_pointer = (uint8_t*)value; \
-    *bytes_pointer = *(ub->iterator + 1); \
-    *(bytes_pointer + 1) = *ub->iterator ; \
-
-#define UCDR_DESERIALIZE_BYTE_4_CORE() \
-    uint8_t* bytes_pointer = (uint8_t*)value; \
-    *bytes_pointer = *(ub->iterator + 3); \
-    *(bytes_pointer + 1) = *(ub->iterator + 2); \
-    *(bytes_pointer + 2) = *(ub->iterator + 1); \
-    *(bytes_pointer + 3) = *ub->iterator;
-
-#define UCDR_DESERIALIZE_BYTE_8_CORE() \
-    uint8_t* bytes_pointer = (uint8_t*)value; \
-    *bytes_pointer = *(ub->iterator + 7); \
-    *(bytes_pointer + 1) = *(ub->iterator + 6); \
-    *(bytes_pointer + 2) = *(ub->iterator + 5); \
-    *(bytes_pointer + 3) = *(ub->iterator + 4); \
-    *(bytes_pointer + 4) = *(ub->iterator + 3); \
-    *(bytes_pointer + 5) = *(ub->iterator + 2); \
-    *(bytes_pointer + 6) = *(ub->iterator + 1); \
-    *(bytes_pointer + 7) = *ub->iterator;
 
 #define UCDR_DESERIALIZE_BYTE_N(TYPE, SIZE, ENDIAN) \
-    ub->iterator += ucdr_buffer_alignment(ub, SIZE); \
-    if(ucdr_check_final_buffer_behavior(ub, SIZE)) \
+    if(ucdr_enough_space(us, SIZE)) \
     { \
         if(UCDR_MACHINE_ENDIANNESS == ENDIAN) \
         { \
-            memcpy((void*)value, ub->iterator, SIZE); \
+            size_t remaining_size = SIZE; \
+            do \
+            { \
+                size_t buffer_available_size = ucdr_buffer_remaining_size(us); \
+                if (remaining_size <= buffer_available_size) \
+                { \
+                    memcpy((uint8_t*)value + (SIZE - remaining_size), us->iterator, remaining_size); \
+                    ucdr_advance_stream(us, remaining_size); \
+                    remaining_size = 0; \
+                } \
+                else \
+                { \
+                    memcpy((uint8_t*)value + (SIZE - remaining_size), us->iterator, buffer_available_size); \
+                    ucdr_promote_buffer(us); \
+                    remaining_size -= buffer_available_size; \
+                } \
+            } while (remaining_size > 0); \
         } \
         else \
         { \
-            UCDR_DESERIALIZE_BYTE_ ## SIZE ## _CORE() \
+            uint8_t* bytes_pointer = (uint8_t*)value; \
+            size_t remaining_size = SIZE; \
+            do \
+            { \
+                size_t buffer_available_size = ucdr_buffer_remaining_size(us); \
+                if (remaining_size <= buffer_available_size) \
+                { \
+                    for (size_t i = (SIZE - remaining_size); i < SIZE; ++i) \
```

Comment:
I like it, the compiler in O3 should be able to produce exactly the same code as the unrolled version and is more maintainable and less error prune

Reply:
:wink: